phoenix-core/src/main/java/org/apache/phoenix/index/GlobalIndexRowVerifier.java

```diff
@@ -0,0 +1,208 @@
+            dataHTable = ServerUtil.getHTableForCoprocessorScan(env,
+                    SchemaUtil.getPhysicalTableName(dataTableName, env.getConfiguration()));
+            if (indexMaintainer == null) {
+                byte[] md = scan.getAttribute(PhoenixIndexCodec.INDEX_PROTO_MD);
+                List<IndexMaintainer> maintainers = IndexMaintainer.deserialize(md, true);
+                Iterator<IndexMaintainer> maintainerIterator = maintainers.iterator();
+                while (maintainerIterator.hasNext()) {
+                    IndexMaintainer maintainer = maintainerIterator.next();
+                    if (Bytes.compareTo(indexTableName, maintainer.getIndexTableName()) == 0) {
+                        indexMaintainer = maintainer;
+                        break;
+                    }
+                }
+            }
+            if (indexMaintainer == null) {
+                throw new DoNotRetryIOException(
+                        "rebuildSingleIndexRow: IndexMaintainer is not included in scan attributes for " +
+                                region.getRegionInfo().getTable().getNameAsString());
+            }
+            if (viewConstants == null) {
+                viewConstants = IndexUtil.deserializeViewConstantsFromScan(scan);
+            }
+            buildIndexScan.setAttribute(BaseScannerRegionObserver.UNGROUPED_AGG, TRUE_BYTES);
+            buildIndexScan.setAttribute(PhoenixIndexCodec.INDEX_PROTO_MD, scan.getAttribute(PhoenixIndexCodec.INDEX_PROTO_MD));
+            buildIndexScan.setAttribute(BaseScannerRegionObserver.REBUILD_INDEXES, TRUE_BYTES);
+        }
+        // Rebuild the index row from the corresponding the row in the the data table
+        byte[] dataRowKey = indexMaintainer.buildDataRowKey(new ImmutableBytesWritable(indexRowKey), viewConstants);
+        buildIndexScan.withStartRow(dataRowKey, true);
+        buildIndexScan.withStopRow(dataRowKey, true);
+        buildIndexScan.setTimestamp(ts);
+        buildIndexScan.setRaw(true);
+        ResultScanner resultScanner = dataHTable.getScanner(buildIndexScan);
+        List<Cell> row = new ArrayList<Cell>();
+        if (resultScanner.next() == null) {
+            // Corresponding row does not exist in data table. Need to delete the row from index if it is old enough
+            if (ts - EnvironmentEdgeManager.currentTimeMillis() > AGE_THRESHOLD_TO_DELETE_MS) {
+                Delete del = new Delete(indexRowKey, ts);
+                Mutation[] mutations = new Mutation[]{del};
+                region.batchMutate(mutations);
+            }
+            return row;
+        }
+
+        // Scan the newly build row
+        indexScan.withStartRow(indexRowKey, true);
+        indexScan.withStopRow(indexRowKey, true);
+        RegionScanner regionScanner = region.getScanner(indexScan);
+        regionScanner.next(row);
+        regionScanner.close();
+        return row;
+    }
+
+    public boolean verifyRow(List<Cell> cellList) throws IOException {
```

Comment:
Name is a bit misleading here -- I would assume a boolean method called "verifyRow" was purely a read-only operation, but if verification fails, the parameter is changed and rebuilt. 

I'm tempted to ask for something like `public List<Cell> getVerifiedCells(List<Cell> cellList)`, with an exception thrown in what's the false case here, though we'd have to be very careful to prevent memory leaks if we went down that route.

Reply:
I will rename it to verifyRowAndReconstructIfNecessary

Reply:
I have restructure the code and addressed this naming issue